AutoCollection/StatsBeat.ts

```diff
@@ -0,0 +1,236 @@
+    private _instrumentations: number = Constants.StatsbeatInstrumentation.NONE;
+
+    constructor(config?: Config) {
+        this._isInitialized = false;
+        this._config = config;
+        this._statsbeatConfig = new Config(STATSBEAT_CONNECTIONSTRING);
+        this._sender = new Sender(this._statsbeatConfig);
+    }
+
+    public enable(isEnabled: boolean) {
+        this._isEnabled = isEnabled;
+        if (this._isEnabled && !this._isInitialized) {
+            this._getCustomProperties();
+            this._isInitialized = true;
+        }
+        if (isEnabled) {
+            if (!this._handle) {
+                this._lastRequests = {
+                    totalRequestCount: this._totalRequestCount,
+                    time: +new Date
+                };
+                this._handle = setInterval(() => {
+                    this.trackStatsbeatMetrics()
+                }, this._collectionInterval);
+                this._handle.unref(); // Allow the app to terminate even while this loop is going on
+            }
+        } else {
+            if (this._handle) {
+                clearInterval(this._handle);
+                this._handle = null;
+            }
+        }
+    }
+
+    public isInitialized() {
+        return this._isInitialized;
+    }
+
+    public isEnabled() {
+        return this._isEnabled;
+    }
+
+    public setCodelessAttach() {
+        this._attach = Constants.StatsbeatAttach.codeless;
+    }
+
+    public addFeature(feature: Constants.StatsbeatFeature) {
+        this._features |= feature;
+    }
+
+    public removeFeature(feature: Constants.StatsbeatFeature) {
+        this._features &= ~feature;
+    }
+
+    public addInstrumentation(instrumentation: Constants.StatsbeatInstrumentation) {
+        this._instrumentations |= instrumentation;
+    }
+
+    public removeInstrumentation(instrumentation: Constants.StatsbeatInstrumentation) {
+        this._instrumentations &= ~instrumentation;
+    }
+
+    public countRequest(duration: number, success: boolean) {
+        if (!this.isEnabled()) {
+            return;
+        }
+        this._totalRequestCount++;
+        this._intervalRequestExecutionTime += duration;
+        if (success === false) {
+            this._totalFailedRequestCount++;
+        }
+        else {
+            this._totalSuccesfulRequestCount++;
+        }
+    }
+
+    public countException() {
+        if (!this.isEnabled()) {
+            return;
+        }
+        this._exceptionCount++;
+    }
+
+    public countThrottle() {
+        if (!this.isEnabled()) {
+            return;
+        }
+        this._throttleCount++;
+    }
+
+    public countRetry() {
+        if (!this.isEnabled()) {
+            return;
+        }
+        this._retryCount++;
+    }
+
+    public trackStatsbeatMetrics() {
+        try {
+            this._trackRequestDuration();
+            this._trackRequestsCount();
+        }
+        catch (error) {
```

Comment:
try / catch doesn't work for promise (async) request responses as when the promise rejects the exception is thrown to an event handler and not this code -- because the JS execution has long since moved on.

Reply:
You would really have to do something like this
```
this._trackRequestDuration().then(() => { }, () => {});
```

Which would effectively ignore the "success" and "reject" responses.